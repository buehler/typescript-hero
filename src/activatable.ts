export default interface Activatable {
  setup(): void;
  start(): void;
  stop(): void;
  dispose(): void;
}
