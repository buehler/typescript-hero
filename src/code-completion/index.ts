import { inject, injectable } from 'inversify';
import { DeclarationInfo, TypescriptParser } from 'typescript-parser';
import {
  CancellationToken,
  commands,
  CompletionItem,
  CompletionItemProvider,
  ExtensionContext,
  languages,
  Position,
  TextDocument,
  workspace,
} from 'vscode';

import Activatable from '../activatable';
import Configuration from '../configuration';
import DeclarationManager from '../declarations/declaration-manager';
import iocSymbols, { ImportManagerProvider } from '../ioc-symbols';
import { Logger } from '../utilities/logger';
import { getDeclarationsFilteredByImports, getItemKind, getScriptKind } from '../utilities/utility-functions';

const REGEX_COMMENT = /^\s*(?:\/\/|\/\*|\*\/|\*)/;

/**
 * Extension that provides code completion for typescript and javascript files.
 * Uses the calculated index to provide information.
 *
 * @export
 * @class CodeCompletion
 * @implements {Activatable}
 * @implements {CompletionItemProvider}
 */
@injectable()
export class CodeCompletion implements Activatable, CompletionItemProvider {

  constructor(
    @inject(iocSymbols.extensionContext) private context: ExtensionContext,
    @inject(iocSymbols.logger) private logger: Logger,
    @inject(iocSymbols.configuration) private config: Configuration,
    @inject(iocSymbols.importManager) private importManagerProvider: ImportManagerProvider,
    @inject(iocSymbols.declarationManager) private declarationManager: DeclarationManager,
    @inject(iocSymbols.parser) private parser: TypescriptParser,
  ) { }

  public setup(): void {
    this.logger.debug('[CodeCompletion] Setting up CodeCompletion.');
    this.logger.debug(
      '[CodeCompletion] Registering for languages.',
      { languages: this.config.parseableLanguages() },
    );

    for (const lang of this.config.parseableLanguages()) {
      this.context.subscriptions.push(languages.registerCompletionItemProvider(lang, this));
    }

    this.context.subscriptions.push(
      commands.registerCommand(
        'typescriptHero.codeCompletion.executeIntellisenseItem',
        (document: TextDocument, declaration: DeclarationInfo) =>
          this.executeIntellisenseItem(document, declaration),
      ),
    );
  }

  public start(): void {
    this.logger.info('[CodeCompletion] Starting up CodeCompletion.');
  }

  public stop(): void {
    this.logger.info('[CodeCompletion] Stopping CodeCompletion.');
  }

  public dispose(): void {
    this.logger.debug('[CodeCompletion] Disposing CodeCompletion.');
  }

  public async provideCompletionItems(
    document: TextDocument,
    position: Position,
    token: CancellationToken,
  ): Promise<CompletionItem[] | null> {
    const index = this.declarationManager.getIndexForFile(document.uri);
    const actualWorkspace = workspace.getWorkspaceFolder(document.uri);

    if (!index || !actualWorkspace || !index.indexReady) {
      this.logger.info('[CodeCompletion] index not ready or no workspace folder selected');
      return null;
    }

    const wordAtPosition = document.getWordRangeAtPosition(position);
    const lineText = document.lineAt(position.line).text;

    let searchWord = '';

    if (wordAtPosition && wordAtPosition.start.character < position.character) {
      const word = document.getText(wordAtPosition);
      searchWord = word.substr(0, position.character - wordAtPosition.start.character);
    }

    if (
      !searchWord ||
      token.isCancellationRequested ||
      (lineText.substring(0, position.character).match(/["'`]/g) || []).length % 2 === 1 ||
      REGEX_COMMENT.test(lineText) ||
      lineText.startsWith('import ') ||
      new RegExp(`(?:\w*\.)+${searchWord}`).test(lineText.substring(0, position.character))
    ) {
      this.logger.debug(
        '[CodeCompletion] did not match criteria to provide intellisense',
        { searchWord, lineText, indexReady: index.indexReady },
      );
      return null;
    }

    this.logger.debug(`[CodeCompletion] provide code completion for "${searchWord}"`);
    const profiler = this.logger.startTimer();

    const parsed = await this.parser.parseSource(document.getText(), getScriptKind(document.fileName));
    const declarations = getDeclarationsFilteredByImports(
      index.declarationInfos,
      document.fileName,
      parsed.imports,
      actualWorkspace.uri.fsPath,
    )
      .filter(o => !parsed.declarations.some(d => d.name === o.declaration.name))
      .filter(o => !parsed.usages.some(d => d === o.declaration.name));

    const items: CompletionItem[] = [];
    for (const declaration of declarations.filter(
      o => o.declaration.name.toLowerCase().indexOf(searchWord.toLowerCase()) >= 0)
    ) {
      const item = new CompletionItem(declaration.declaration.name, getItemKind(declaration.declaration));

      item.detail = declaration.from;
      item.command = {
        arguments: [document, declaration],
        title: 'Execute intellisense insert',
        command: 'typescriptHero.codeCompletion.executeIntellisenseItem',
      };
      if (this.config.codeCompletion.completionSortMode(document.uri) === 'bottom') {
        item.sortText = `9999-${declaration.declaration.name}`;
      }
      items.push(item);
    }

    profiler.done({ message: `[CodeCompletion] calculated code completions` });
    return items;
  }

  /**
   * Executes a intellisense item that provided a document and a declaration to add.
   * Does make the calculation of the text edits async.
   *
   * @private
   * @param {TextDocument} document
   * @param {DeclarationInfo} declaration
   * @returns {Promise<void>}
   * @memberof CodeCompletion
   */
  private async executeIntellisenseItem(document: TextDocument, declaration: DeclarationInfo): Promise<void> {
    this.logger.info(
      '[CodeCompletion] Execute code completion action.',
      { specifier: declaration.declaration.name, library: declaration.from },
    );
    await (await this.importManagerProvider(document)).addDeclarationImport(declaration).commit();
  }
}
