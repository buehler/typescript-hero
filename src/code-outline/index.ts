import { inject, injectable } from 'inversify';
import { Subscription } from 'rxjs';
import { File, TypescriptParser } from 'typescript-parser';
import {
  Disposable,
  Event,
  EventEmitter,
  ExtensionContext,
  ProviderResult,
  TreeDataProvider,
  window,
  workspace,
} from 'vscode';

import Activatable from '../activatable';
import Configuration from '../configuration';
import iocSymbols from '../ioc-symbols';
import { Logger } from '../utilities/Logger';
import { getScriptKind } from '../utilities/utilityFunctions';
import BaseStructureTreeItem from './base-structure-tree-item';
import DeclarationStructureTreeItem from './declaration-structure-tree-item';
import DisabledStructureTreeItem from './disabled-structure-tree-item';
import { ImportsStructureTreeItem } from './imports-structure-tree-item';
import NotParseableStructureTreeItem from './not-parseable-structure-tree-item';
import ResourceStructureTreeItem from './resource-structure-tree-item';

@injectable()
export default class CodeOutline implements Activatable, TreeDataProvider<BaseStructureTreeItem> {
  private _onDidChangeTreeData: EventEmitter<BaseStructureTreeItem | undefined>;

  private subscription: Subscription;
  private disposables: Disposable[] = [];
  private documentCache?: File;

  public get onDidChangeTreeData(): Event<BaseStructureTreeItem | undefined> {
    return this._onDidChangeTreeData.event;
  }

  constructor(
    @inject(iocSymbols.extensionContext) private context: ExtensionContext,
    @inject(iocSymbols.logger) private logger: Logger,
    @inject(iocSymbols.configuration) private config: Configuration,
    @inject(iocSymbols.parser) private parser: TypescriptParser,
  ) { }

  public setup(): void {
    this.logger.debug('Setting up DocumentOutline.');
    this.subscription = this.config.configurationChanged.subscribe(() => {
      if (this.config.codeOutline.isEnabled() && !this.disposables) {
        this.start();
      } else if (!this.config.codeOutline.isEnabled() && this.disposables) {
        this.stop();
      }
    });
  }

  public start(): void {
    if (!this.config.codeOutline.isEnabled()) {
      this.logger.debug(`Not starting DocumentOutline. It's disabled by config.`);
      return;
    }
    this.logger.debug('Starting up DocumentOutline.');
    this._onDidChangeTreeData = new EventEmitter<BaseStructureTreeItem | undefined>();
    this.disposables.push(window.registerTreeDataProvider('codeOutline', this));
    this.context.subscriptions.push(this._onDidChangeTreeData);
    this.context.subscriptions.push(window.onDidChangeActiveTextEditor(() => this.activeWindowChanged()));
    this.context.subscriptions.push(workspace.onDidSaveTextDocument(() => this.activeWindowChanged()));
  }

  public stop(): void {
    if (this.config.codeOutline.isEnabled()) {
      this.logger.debug(`Not stopping DocumentOutline. It's enabled by config.`);
      return;
    }
    this.logger.debug('Stopping DocumentOutline.');
    for (const disposable of this.disposables) {
      disposable.dispose();
    }
    this.disposables = [];
  }

  public dispose(): void {
    this.logger.debug('Disposing DocumentOutline.');
    if (this.subscription) {
      this.subscription.unsubscribe();
      delete this.subscription;
    }
    for (const disposable of this.disposables) {
      disposable.dispose();
    }
    this.disposables = [];
  }

  public getTreeItem(element: BaseStructureTreeItem): BaseStructureTreeItem {
    return element;
  }

  public async getChildren(element?: BaseStructureTreeItem): Promise<ProviderResult<BaseStructureTreeItem[]>> {
    if (!window.activeTextEditor) {
      return [];
    }

    if (!this.config.codeOutline.isEnabled()) {
      return [new DisabledStructureTreeItem()];
    }

    if (!this.config.parseableLanguages().some(
      lang => lang === window.activeTextEditor!.document.languageId,
    )) {
      return [new NotParseableStructureTreeItem()];
    }

    if (!this.documentCache) {
      try {
        this.documentCache = await this.parser.parseSource(
          window.activeTextEditor.document.getText(),
          getScriptKind(window.activeTextEditor.document.fileName),
        );
      } catch (e) {
        this.logger.error(
          `[DocumentOutline] document could not be parsed, error: ${e}`,
        );
        return [];
      }
    }

    if (!element) {
      const items: BaseStructureTreeItem[] = [];
      if (this.documentCache.imports && this.documentCache.imports.length) {
        items.push(new ImportsStructureTreeItem(this.documentCache, this.context));
      }
      items.push(...this.documentCache.resources.map(r => new ResourceStructureTreeItem(r, this.context)));
      items.push(...this.documentCache.declarations.map(d => new DeclarationStructureTreeItem(d, this.context)));
      return items;
    }
    return element.getChildren();
  }

  /**
     * Method that recalculates the current document when the active window changed.
     *
     * @private
     *
     * @memberof DocumentSymbolStructureExtension
     */
  private activeWindowChanged(): void {
    this.logger.debug('[DocumentOutline] activeWindowChanged, reparsing');
    this.documentCache = undefined;
    this._onDidChangeTreeData.fire();
  }
}
