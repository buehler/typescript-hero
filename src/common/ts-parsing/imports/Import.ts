import { Generatable } from '../ts-generation';
import { Clonable } from '../utilities';
import { Node } from './Node';

/**
 * Basic import interface. Defines an import in a document.
 * If no start and end points are given, the import is considered "new".
 * 
 * @export
 * @interface Import
 */
export interface Import extends Clonable, Node, Generatable {
    /**
     * Indicates if the given import is "new" or a parsed one.
     * 
     * @type {boolean}
     * @memberOf Import
     */
    readonly isNew: boolean;

    /**
     * The library name of the given import (This actually is the part after the import statement).
     *
     * @example "express"
     *
     * @type {string}
     * @memberOf Import
     */
    libraryName: string;
}

/**
 * Basic interface for aliased imports. Defines an alias for namespace imports and other aliased imports.
 * 
 * @export
 * @interface AliasedImport
 * @extends {Import}
 */
export interface AliasedImport extends Import {
    /**
     * Alias for the given import. E.g. for a "* as ..." import.
     * 
     * @type {string}
     * @memberOf AliasedImport
     */
    alias: string;
}
