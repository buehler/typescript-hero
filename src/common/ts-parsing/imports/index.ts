export * from './DefaultImport';
export * from './ExternalModuleImport';
export * from './Import';
export * from './NamedImport';
export * from './NamespaceImport';
export * from './StringImport';
