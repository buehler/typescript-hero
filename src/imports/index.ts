export { ImportAppender } from './import-appender';
export { ImportOrganizer } from './import-organizer';
