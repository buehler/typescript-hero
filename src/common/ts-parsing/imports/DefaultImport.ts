import { GenerationOptions } from '../../ts-generation';
import { AliasedImport, importRange } from './Import';
import { Serializable } from 'ts-json-serializer';
import { Range, TextDocument } from 'vscode-languageserver-types';

/**
 * Default import. Imports the default exports of a file.
 * (i.e. import foobar from ...).
 * 
 * @export
 * @class DefaultImport
 * @implements {AliasedImport}
 */
@Serializable({ factory: json => new DefaultImport(json.libraryName, json.alias, json.start, json.end) })
export class DefaultImport implements AliasedImport {
    public get isNew(): boolean {
        return this.start !== undefined && this.end !== undefined;
    }

    constructor(public libraryName: string, public alias: string, public start?: number, public end?: number) { }

    /**
     * Generates typescript code out of the actual import.
     * 
     * @param {GenerationOptions} {stringQuoteStyle, eol}
     * @returns {string}
     * 
     * @memberOf DefaultImport
     */
    public generateTypescript({stringQuoteStyle, eol}: GenerationOptions): string {
        return `import ${this.alias} from ${stringQuoteStyle}${this.libraryName}${stringQuoteStyle}${eol}\n`;
    }

    /**
     * Calculates the document range of the node in the given document.
     * 
     * @param {TextDocument} document
     * @returns {Range}
     * 
     * @memberOf DefaultImport
     */
    public getRange(document: TextDocument): Range {
        return importRange(document, this.start, this.end);
    }

    /**
     * Clone the current import object.
     * 
     * @returns {DefaultImport}
     * 
     * @memberOf DefaultImport
     */
    public clone(): DefaultImport {
        return new DefaultImport(this.libraryName, this.alias, this.start, this.end);
    }
}
