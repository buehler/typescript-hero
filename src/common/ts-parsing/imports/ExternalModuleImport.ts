import { GenerationOptions } from '../../ts-generation';
import { DocumentLike } from '../Node';
import { AliasedImport, importRange } from './Import';
import { Serializable } from 'ts-json-serializer';
import { Range } from 'vscode-languageserver-types';

/**
 * Alternative to the namespace import. Can be used by various libraries.
 * (i.e. import foobar = require('foobar')).
 * 
 * @export
 * @class ExternalModuleImport
 * @implements {AliasedImport}
 */
@Serializable({ factory: json => new ExternalModuleImport(json.libraryName, json.alias, json.start, json.end) })
export class ExternalModuleImport implements AliasedImport {
    public get isNew(): boolean {
        return this.start !== undefined && this.end !== undefined;
    }

    constructor(public libraryName: string, public alias: string, public start?: number, public end?: number) { }

    /**
     * Generates typescript code out of the actual import.
     * 
     * @param {GenerationOptions} {stringQuoteStyle, eol}
     * @returns {string}
     * 
     * @memberOf ExternalModuleImport
     */
    public generateTypescript({stringQuoteStyle, eol}: GenerationOptions): string {
        return `import ${this.alias} = require(${stringQuoteStyle}${this.libraryName}${stringQuoteStyle})${eol}\n`;
    }

    /**
     * Calculates the document range of the node in the given document.
     * 
     * @param {DocumentLike} document
     * @returns {Range}
     * 
     * @memberOf ExternalModuleImport
     */
    public getRange(document: DocumentLike): Range {
        return importRange(document, this.start, this.end);
    }

    /**
     * Clone the current import object.
     * 
     * @returns {ExternalModuleImport}
     * 
     * @memberOf ExternalModuleImport
     */
    public clone(): ExternalModuleImport {
        return new ExternalModuleImport(this.libraryName, this.alias, this.start, this.end);
    }
}
