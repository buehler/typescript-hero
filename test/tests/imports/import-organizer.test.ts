import { join } from 'path';
import { Position, Range, TextDocument, Uri, window, workspace } from 'vscode';

import { ImportOrganizer } from '../../../src/imports';
import ioc from '../../../src/ioc';
import iocSymbols from '../../../src/ioc-symbols';
import { expect } from '../setup';

describe('ImportOrganizer', () => {

  describe('import-organizer-file.ts', () => {

    const rootPath = workspace.workspaceFolders![0].uri.fsPath;
    const file = Uri.file(join(rootPath, 'imports', 'import-organizer-file.ts'));
    let document: TextDocument;
    let extension: any;

    before(async () => {
      document = await workspace.openTextDocument(file);
      await window.showTextDocument(document);

      extension = new ImportOrganizer(
        ioc.get(iocSymbols.extensionContext),
        ioc.get(iocSymbols.logger),
        ioc.get(iocSymbols.configuration),
        ioc.get(iocSymbols.importManager),
      );
    });

    afterEach(async () => {
      await window.activeTextEditor!.edit((builder) => {
        builder.delete(new Range(
          new Position(0, 0),
          document.lineAt(document.lineCount - 1).rangeIncludingLineBreak.end,
        ));
      });
    });

    it('should not remove directly exported imports', async () => {
      await window.activeTextEditor!.edit((builder) => {
        builder.insert(
          new Position(0, 0),
          `import * as Foobar from './lol';
import * as Barbaz from './foo';

export { Foobar, Barbaz }
`,
        );
      });

      expect(window.activeTextEditor!.document.getText()).to.matchSnapshot();
      await extension.organizeImports();
      expect(window.activeTextEditor!.document.getText()).to.matchSnapshot();
    });

    it('should not remove directly exported default imports', async () => {
      await window.activeTextEditor!.edit((builder) => {
        builder.insert(
          new Position(0, 0),
          `import Barbaz from './foo';

export { Barbaz }
`,
        );
      });

      expect(window.activeTextEditor!.document.getText()).to.matchSnapshot();
      await extension.organizeImports();
      expect(window.activeTextEditor!.document.getText()).to.matchSnapshot();
    });

    it('should not remove default exported default imports', async () => {
      await window.activeTextEditor!.edit((builder) => {
        builder.insert(
          new Position(0, 0),
          `import Barbaz from './foo';

export default Barbaz;
`,
        );
      });

      expect(window.activeTextEditor!.document.getText()).to.matchSnapshot();
      await extension.organizeImports();
      expect(window.activeTextEditor!.document.getText()).to.matchSnapshot();
    });

  });

});
