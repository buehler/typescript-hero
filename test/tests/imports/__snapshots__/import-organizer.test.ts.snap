// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ImportOrganizer import-organizer-file.ts should not remove default exported default imports 1`] = `
"import Barbaz from './foo';

export default Barbaz;
"
`;

exports[`ImportOrganizer import-organizer-file.ts should not remove default exported default imports 2`] = `
"import Barbaz from './foo';

export default Barbaz;
"
`;

exports[`ImportOrganizer import-organizer-file.ts should not remove directly exported default imports 1`] = `
"import Barbaz from './foo';

export { Barbaz }
"
`;

exports[`ImportOrganizer import-organizer-file.ts should not remove directly exported default imports 2`] = `
"import Barbaz from './foo';

export { Barbaz }
"
`;

exports[`ImportOrganizer import-organizer-file.ts should not remove directly exported imports 1`] = `
"import * as Foobar from './lol';
import * as Barbaz from './foo';

export { Foobar, Barbaz }
"
`;

exports[`ImportOrganizer import-organizer-file.ts should not remove directly exported imports 2`] = `
"import * as Barbaz from './foo';
import * as Foobar from './lol';

export { Foobar, Barbaz }
"
`;

exports[`ImportOrganizer import-organizer-file.tsx should not remove function that is used in tsx 1`] = `
"import { f } from './somewhere';
import * as React from 'react';

export class Comp extends React.Component {
  render() {
    return (
      <div>{f()}</div>
    );
  }
}
"
`;

exports[`ImportOrganizer import-organizer-file.tsx should not remove function that is used in tsx 2`] = `
"import * as React from 'react';

import { f } from './somewhere';

export class Comp extends React.Component {
  render() {
    return (
      <div>{f()}</div>
    );
  }
}
"
`;
